### PR TITLE
feat(alerts): Use metrics layer for snql generation [TET-597]

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at using for `Dataset.Metrics` and `Dataset.PerformanceMetrics` the metrics layer core component `SnubaQueryBuilder`. This condition is defined here: https://github.com/getsentry/sentry/blob/0b7e308ae98ed9320ca6a6f965876ba16a368b5b/src/sentry/snuba/entity_subscription.py#L337-L337

The usage of the metrics layer in this case means that we will use the layer for both snql generation through `get_snql_query` but also for query execution through `run_query`.

The idea behind this PR is that we want to use the `AlertMetricsQueryBuilder` to generate intermediate snql which will be fed into the `mqb_query_transformer` to generate to `MetricsQuery` and at the end this will be transformed by the `SnubaQueryBuilder` back to snql. This snql will be used by alerts to generate a snuba subscription. _Currently this implementation is violating the purpose of the metrics layer, thus in the future we would like to directly support subscriptions in the layer itself._

This PR has touched the following files:
* Overriden the `run_query` and `get_snql_query` in the `AlertMetricsQueryBuilder` in order to support the custom execution without touching the `MetricsQueryBuilder`
* Redefined `uniq`, `uniqIf` and `sumIf` in https://github.com/getsentry/sentry/blob/2b6c452663b78ff752ec2c295ab41eea91775b45/src/sentry/search/events/datasets/metrics_layer.py#L59-L59
* Added new operators to the metrics layer namely `sum_if_column` and `uniq_if_column` which are used by the alerts entity subscriptions.
* Maintained backwards compatibility with the `use_metrics_layer` flag, so that we can switch it to off and the previous logic will be used.

_In order for the query to be properly executed we needed to redefine the `run_query` method so that it can use the snql query generated by the `MetricsQueryBuilder` instead of the one generated by the `AlertMetricsQueryBuilder`, this is because the `run_query` uses inside the transformer that accepts a special snql and not the snql generated by itself, which would be returned by calling the `get_snql_query` method overridden in `AlertMetricsQueryBuilder`._

### CODE REVIEW

File: src/sentry/snuba/entity_subscription.py
Comment (wedamija) [inline src/sentry/snuba/entity_subscription.py:306]: Nit: Since all subclasses of `BaseMetricsEntitySubscription` use one of these two datasets, not totally sure if this is necessary?

Could simplify a lot of the code below though.

**Review by wedamija — COMMENTED** [review]
I've reviewed the code specifically around alerts here + tests
 
All look good to me. I'm probably not the best person to be reviewing metrics layer changes or query builder changes, so will leave it to someone else to give final approval there. But since your local end to end tests look good I think that we're probably good to go.

File: src/sentry/snuba/entity_subscription.py
Comment (iambriccardo) [reply src/sentry/snuba/entity_subscription.py:306]: The goal of my changes was to be as explicit as possible with the code but I can for sure simplify some parts.

Comment (iambriccardo) [reply src/sentry/snuba/entity_subscription.py:306]: I will iterate on some improvements now, since you reviewed the general implementation. Will update the PR!

**Comment by iambriccardo** [comment]
> I've reviewed the code specifically around alerts here + tests
> 
> All look good to me. I'm probably not the best person to be reviewing metrics layer changes or query builder changes, so will leave it to someone else to give final approval there. But since your local end to end tests look good I think that we're probably good to go.

After @wmak is back from vacation I will ask him for a review because I have touched code he owns.

File: src/sentry/snuba/entity_subscription.py
Comment (iambriccardo) [reply src/sentry/snuba/entity_subscription.py:306]: Btw, my ideas of introducing this boolean was to have both implementations that can be switched in case of issues or need. If we decide to avoid completely having a flag, I could simplify the code and move everything to metrics layer.

File: src/sentry/search/events/builder/metrics.py
Comment (wmak) [inline src/sentry/search/events/builder/metrics.py:631]: I'd rather we didn't move this to a helper function, since eventually run_query will just be calling this code directly once we're 100% on the metrics_layer

File: src/sentry/search/events/builder/metrics.py
Comment (wmak) [inline src/sentry/search/events/builder/metrics.py:809]: Isn't this only true because you're calling `super().get_snql_query()`? Wouldn't you not have to do this at all if you didn't overwrite `run_query` since you're overwriting `get_snql_query` already?

To add to this, do we need the `_get_snql_using_metrics_layer` helper function? Can't we just overwrite get_snql_query directly instead?

File: src/sentry/search/events/builder/metrics.py
Comment (wmak) [inline src/sentry/search/events/builder/metrics.py:795]: This is no longer a valid error message in this case

File: tests/snuba/api/endpoints/test_organization_events_mep.py
Comment (wmak) [inline tests/snuba/api/endpoints/test_organization_events_mep.py:448]: Is this intentional? use-metrics-layer is enabled by the `OrganizationEventsMetricsEnhancedPerformanceEndpointTestWithMetricLayer` tests

**Review by wmak — COMMENTED** [review]
I'm locally seeing `tests/snuba/api/endpoints/test_organization_events_mep.py::OrganizationEventsMetricsEnhancedPerformanceEndpointTestWithMetricLayer::test_environment_query` fail

Do we need to add tests to `tests/sentry/snuba/test_entity_subscriptions.py` that include enabling `use_metrics_layer`?

Also just to be a bit more explicit with what I'm looking for here's a diff that's roughly what i mean for the `MetricsQueryBuilder`
https://gist.github.com/wmak/18bea8802be2a4e3258bdd5b2941381d

File: src/sentry/search/events/builder/metrics.py
Comment (iambriccardo) [reply src/sentry/search/events/builder/metrics.py:631]: This has been explicitly done because in the `AlertMetricsQueryBuilder.run_query()` method I have to get the snql for the transformer and run it with the metrics layer. If I would have called `super.run_query()` directly what would have happened is that via the MRO (https://github.com/getsentry/sentry/blob/426295d4aee62db35c2331f23c5b5e1e2583cd90/src/sentry/search/events/builder/metrics.py#L634-L636) we would have had the snql being generated by the `AlertMetricsQueryBuilder.get_snql_query()` but that snql is already out of `SnubaQueryBuilder`, thus it is not understood by the transformer.

TL;DR
I have done this in order to avoid duplication of the method.

File: tests/snuba/api/endpoints/test_organization_events_mep.py
Comment (iambriccardo) [reply tests/snuba/api/endpoints/test_organization_events_mep.py:448]: This is my bad, i tend to hardcode it for doing some tests quicker, will remove it.

File: src/sentry/search/events/builder/metrics.py
Comment (iambriccardo) [reply src/sentry/search/events/builder/metrics.py:809]: We could do that sure, I just wanted to reduce code duplication.

File: src/sentry/search/events/builder/metrics.py
Comment (iambriccardo) [reply src/sentry/search/events/builder/metrics.py:795]: You are right, will make it more meaningful!

File: src/sentry/search/events/builder/metrics.py
Comment (iambriccardo) [reply src/sentry/search/events/builder/metrics.py:809]: In your diff, you remove the `run_query` function which will mean we are going to use the `run_query` of the parent class that will use the `get_snql_query` of the child class during execution. All of my code was to circumvent this issue.

Comment (wmak) [reply src/sentry/search/events/builder/metrics.py:809]: If i understand what you're trying to do better this may be a more accurate diff for what you're doing: https://gist.github.com/wmak/d654c19be1fb1304a340f9310b0ca115

The idea is to introduce a new `get_metrics_layer_query` that returns a Request compatible with the metric layer, return `NotImplemented` on the base `MetricsQueryBuilder.get_snql_query`. And only implement `get_snql_query` on the `AlertQueryBuilder`

Comment (iambriccardo) [reply src/sentry/search/events/builder/metrics.py:809]: This approach could work but we then would simply need to implement`get_metrics_layer_query` in the `MetricsQueryBuilder` as that building process is not different from the alerts one and this would converge to an implementation similar to mine.

File: src/sentry/search/events/builder/metrics.py
Comment (wmak) [inline src/sentry/search/events/builder/metrics.py:821]: nit, can this be a docstring instead?
```suggestion
        """
        We are overriding this method here because in the case of alerts, we would like to use the snql query
        generated by the metrics layer. This implementation goes against the rationale of the metrics layer and should
        be removed as soon as snuba subscriptions will be supported by the layer.
        The logic behind this method is that MetricsQueryBuilder will generate a dialect of snql via the
        "get_metrics_layer_snql_query()" method, which will be fed into the transformer that will generate
        the MetricsQuery which is the DSL of querying that the metrics layer understands. Once this is generated,
        we are going to import the purposfully hidden SnubaQueryBuilder which is a component that takes a MetricsQuery
        """
```

File: src/sentry/search/events/builder/metrics.py
Comment (wmak) [inline src/sentry/search/events/builder/metrics.py:840]: nit:
```suggestion
            if len(snuba_queries) != 1:
```

**Review by wmak — COMMENTED** [review]
Changes to the QueryBuilder look good to me now, ~my biggest remaining concern is still if we need to add tests to `tests/sentry/snuba/test_entity_subscriptions.py` that include enabling `use_metrics_layer`?~

Nevermind misunderstood the code tests cause I was expecting to see something where the feature flag would be enabled. As per our slack conversation, we should update the entity code so it uses the metric_layer based on the existing `use-metrics-layer` feature flag

**Review by wmak — APPROVED** [review]
Approved, but please don't merge without a feature flag in the entity subscription code

**Pull request merged**
